flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/InternalTimerService.java

```diff
@@ -18,43 +18,306 @@
 	 * Deletes the timer for the given key and namespace.
 	 */
-	void deleteEventTimeTimer(N namespace, long time);
+	abstract public void deleteEventTimeTimer(N namespace, long time);
+
+	/**
+	 * Returns the timers for the given key group.
+	 */
+	abstract public Set<InternalTimer<K, N>> getEventTimeTimersForKeyGroup(int keyGroup);
+
+	/**
+	 * Returns the timers for the given key group.
+	 */
+	abstract public Set<InternalTimer<K, N>> getProcessingTimeTimersForKeyGroup(int keyGroup);
+
+	/**
+	 * Restores the timers for the given key group.
+	 */
+	abstract public void restoreEventTimeTimersForKeyGroup(int keyGroup, Iterable<InternalTimer<K, N>> timers);
+
+	/**
+	 * Restores the timers for the given key group.
+	 */
+	abstract public void restoreProcessingTimeTimersForKeyGroup(int keyGroup, Iterable<InternalTimer<K, N>> timers);
+
+	/**
+	 * Starts the execution of the timer service
+	 */
+	abstract public void start();
+
+	/**
+	 * Closes the timer service.
+	 */
+	abstract public void close();
+	
+	public void advanceWatermark(long watermark) throws Exception {
+		if (watermark < currentWatermark) {
+			throw new IllegalStateException("The watermark is late.");
+		}
+		
+		currentWatermark = watermark;
+		
+		onEventTime(watermark);
+	}
+
+	/**
+	 * Snapshots the timers (both processing and event time ones) for a given {@code keyGroupIdx}.
+	 * @param stream the stream to write to.
+	 * @param keyGroupIdx the id of the key-group to be put in the snapshot.
+	 */
+	public void snapshotTimersForKeyGroup(DataOutputViewStreamWrapper stream, int keyGroupIdx) throws Exception {
+		InstantiationUtil.serializeObject(stream, keySerializer);
+		InstantiationUtil.serializeObject(stream, namespaceSerializer);
+
+		// write the event time timers
+		Collection<InternalTimer<K, N>> eventTimers = getEventTimeTimersForKeyGroup(keyGroupIdx);
+		if (eventTimers != null) {
+			stream.writeInt(eventTimers.size());
+			for (InternalTimer<K, N> timer : eventTimers) {
+				this.timerSerializer.serialize(timer, stream);
+			}
+		} else {
+			stream.writeInt(0);
+		}
+
+		// write the processing time timers
+		Collection<InternalTimer<K, N>> processingTimers = getProcessingTimeTimersForKeyGroup(keyGroupIdx);
+		if (processingTimers != null) {
+			stream.writeInt(processingTimers.size());
+			for (InternalTimer<K, N> timer : processingTimers) {
+				this.timerSerializer.serialize(timer, stream);
+			}
+		} else {
+			stream.writeInt(0);
+		}
+	}
+
+	/**
+	 * Restore the timers (both processing and event time ones) for a given {@code keyGroupIdx}.
+	 * @param stream the stream to read from.
+	 * @param keyGroupIdx the id of the key-group to be put in the snapshot.
+	 * @param userCodeClassLoader the class loader that will be used to deserialize
+	 * 								the local key and namespace serializers.
+	 */
+	public void restoreTimersForKeyGroup(DataInputViewStreamWrapper stream, int keyGroupIdx, ClassLoader userCodeClassLoader) throws IOException, ClassNotFoundException {
```

Comment:
Same comments as for snapshot apply here: we can reduce visibility and refactor to de-duplicate code.